codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -135,123 +139,183 @@ public void close() { }
                     }
-                } while (remainingData != 0 || padding != 0);
-            }
+                }
+                // Write the frame padding.
+                if (paddingBytes(framePaddingBytes) > 0) {
+                    ctx.write(ZERO_BUFFER.slice(0, paddingBytes(framePaddingBytes)),
+                              promiseAggregator.newPromise());
+                }
+            } while (remainingData != 0 || padding != 0);
+            return promiseAggregator.doneAllocatingPromises();
         } catch (Throwable cause) {
-            if (frameHeader != null) {
+            return handleFailure(data, frameHeader, promise, cause, promiseAggregator);
+        }
+    }
+
+    private ChannelFuture writeDataNoPadding(ChannelHandlerContext ctx, int streamId, ByteBuf data,
+                                             boolean endStream, ChannelPromise promise) {
+        int dataLength = data.readableBytes();
+        if (dataLength <= maxFrameSize && (data.hasArray() || dataLength <= COPY_SLICE_CUTOFF_BYTES)) {
+            return writeCopySingleFrameDataNoPadding(ctx, streamId, data, endStream, promise);
+        }
+        return writeSlicedDataNoPadding(ctx, streamId, data, endStream, promise);
+    }
+
+    private ChannelFuture writeSlicedDataNoPadding(ChannelHandlerContext ctx, int streamId, ByteBuf data,
+                                                   boolean endStream, ChannelPromise promise) {
+        SimpleChannelPromiseAggregator promiseAggregator =
+                new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+        ByteBuf frameHeader = null;
+        try {
+            int remainingData = data.readableBytes();
+            final Http2Flags flags = new Http2Flags();
+            flags.endOfStream(false);
+            flags.paddingPresent(false);
+            // Fast path to write frames of payload size maxFrameSize first.
+            if (remainingData > maxFrameSize) {
+                frameHeader = ctx.alloc().buffer(FRAME_HEADER_LENGTH);
+                remainingData = writeDataFrames(ctx, streamId, data, frameHeader,
+                                                flags, promiseAggregator, remainingData, maxFrameSize);
                 frameHeader.release();
+                frameHeader = null;
             }
-            // Use a try/finally here in case the data has been released before calling this method. This is not
-            // necessary above because we internally allocate frameHeader.
+            // Write the header.
+            final ByteBuf frameHeader2 = ctx.alloc().buffer(FRAME_HEADER_LENGTH);
+            flags.endOfStream(endStream);
+            writeFrameHeaderInternal(frameHeader2, remainingData, DATA, flags, streamId);
+            ctx.write(frameHeader2, promiseAggregator.newPromise());
+
+            // Write the payload.
+            ByteBuf lastFrame = data.readSlice(remainingData);
```

Comment:
We could also just adjust indexes and so reduce object creation. That said maybe we should just keep it as it is as it makes things a bit easier.
```suggestion
            ByteBuf lastFrame = data.writerIndex(data.readerIndex + remainingData);
```

Reply:
@normanmaurer I've accepted the changes!